### DISCUSSION TITLE
Not getting actual key-value in the command_results table

### DISCUSSION THREAD

**Discussion by ashishmeher216** [micromdm/nanomdm#119] [Q&A] [answered]
Let's say I install a profile which blocks the screenshot using allowScreenShot as false. The profile is installed on the device successfully and is working as expected.

My profile is :

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
    <key>PayloadContent</key>
    <array>
        <dict>
	    <key>allowScreenShot</key>
            <false/>
            <key>PayloadIdentifier</key>
            <string>com.miniorange.mdm.restriction_payload_3</string>
            <key>PayloadType</key>
            <string>com.apple.applicationaccess</string>
            <key>PayloadUUID</key>
            <string>53bec1be-ffec-4f88-acbd-b02aee8f04a11</string>
            <key>PayloadVersion</key>
            <integer>1</integer>
        </dict>
    </array>
    <key>PayloadDisplayName</key>
    <string>Restrictions</string>
    <key>PayloadIdentifier</key>
    <string>com.miniorange.mdm.conf_profile_3</string>
    <key>PayloadType</key>
    <string>Configuration</string>
    <key>PayloadUUID</key>
    <string>6020206c-12c2-4ada-987a-dd4c560ca73c</string>
    <key>PayloadVersion</key>
    <integer>1</integer>
</dict>
</plist>
`

But when I query the device for ProfileList using the cmdr.py tool, I get the list of profiles but it does not mention anything about the allowScreenShot. It contained the Profile Payload though.

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>CommandUUID</key>
	<string>3c62d4ab-d296-498a-afaf-1ab89083f49b</string>
	<key>ProfileList</key>
	<array>
		<dict>
			<key>HasRemovalPasscode</key>
			<false/>
			<key>IsEncrypted</key>
			<false/>
			<key>IsManaged</key>
			<true/>
			<key>PayloadContent</key>
			<array>
				<dict>
					<key>PayloadIdentifier</key>
					<string>com.miniorange.mdm.restriction_payload_3</string>
					<key>PayloadType</key>
					<string>com.apple.applicationaccess</string>
					<key>PayloadUUID</key>
					<string>53bec1be-ffec-4f88-acbd-b02aee8f04a11</string>
					<key>PayloadVersion</key>
					<integer>1</integer>
				</dict>
			</array>
			<key>PayloadDisplayName</key>
			<string>Restrictions</string>
			<key>PayloadIdentifier</key>
			<string>com.miniorange.mdm.conf_profile_3</string>
			<key>PayloadRemovalDisallowed</key>
			<false/>
			<key>PayloadUUID</key>
			<string>6020206c-12c2-4ada-987a-dd4c560ca73c</string>
			<key>PayloadVersion</key>
			<integer>1</integer>
		</dict>
		<dict>
			<key>HasRemovalPasscode</key>
			<false/>
			<key>IsEncrypted</key>
			<false/>
			<key>IsManaged</key>
			<true/>
			<key>PayloadContent</key>
			<array>
				<dict>
					<key>PayloadIdentifier</key>
					<string>com.github.micromdm.scep</string>
					<key>PayloadType</key>
					<string>com.apple.security.scep</string>
					<key>PayloadUUID</key>
					<string>CB90E976-AD44-4B69-8108-8095E6260978</string>
					<key>PayloadVersion</key>
					<integer>1</integer>
				</dict>
				<dict>
					<key>PayloadIdentifier</key>
					<string>com.github.micromdm.nanomdm.mdm</string>
					<key>PayloadType</key>
					<string>com.apple.mdm</string>
					<key>PayloadUUID</key>
					<string>96B11019-B54C-49DC-9480-43525834DE7B</string>
					<key>PayloadVersion</key>
					<integer>1</integer>
				</dict>
			</array>
			<key>PayloadDisplayName</key>
			<string>Enrollment Profile</string>
			<key>PayloadIdentifier</key>
			<string>com.github.micromdm.nanomdm</string>
			<key>PayloadRemovalDisallowed</key>
			<false/>
			<key>PayloadUUID</key>
			<string>F9760DD4-F2D1-4F29-8D2C-48D52DD0A9B3</string>
			<key>PayloadVersion</key>
			<integer>1</integer>
		</dict>
	</array>
	<key>Status</key>
	<string>Acknowledged</string>
	<key>UDID</key>
	<string>00008110-001970203698401E</string>
</dict>
</plist>

What am I missing?

**Comment by jessepeterson** [comment, marked as answer]
This is expected. As you can see in the [ProfileList response payload](https://developer.apple.com/documentation/devicemanagement/profilelistresponse/profilelistitem/payloadcontentitem) Apple devices only return the top-level information on the payloads installed. You do not get the contents (i.e. the 'full' or 'raw' profiles) returned.

**Reply by ashishmeher216** [reply]
So, how do we get the actual restrictions applied on a device? We have to store the payload details of the profile on our end. Is there any other workaround?

**Reply by jessepeterson** [reply]
I'm confused. You send the profiles to the device. Presumably you know what you sent? :) Or am I misunderstanding?

I think typically MDMs have a store of profiles that are identified by profile UUID & identifier. You then query the device to see if that/those profiles are installed (by the UUID & id) and then install (or update) them if they are not current or not present. This is the workflow that the [NanoCMD profile workflow](https://github.com/micromdm/nanocmd/blob/main/docs/operations-guide.md#profile-workflow) takes, for example.

**Reply by ashishmeher216** [reply]
So, I have to store all the profiles. Let's say I have multiple  profiles, one to disable screenshot only, one to prevent copy/paste from managed pasteboard, etc. I have to store all these profiles somewhere (maybe database). I can distribute them to multiple devices. And when I query about ProfileList in a particular device, I have to compare the **PayloadIdentifier** I get from Apple with my stored profiles and get the details of the profile from there.

Am I correct?

**Reply by jessepeterson** [reply]
Yeah, that sounds about right. :) Note you probably also want to compare the `PayloadUUID` as well. Depending on if you manage macOS or iOS/iPadOS/tvOS, etc. they have slightly different semantics.

You _can_ dynamically generate them, too. But as you said you'll need to keep track of the `PayloadIdentifier` and `PayloadUUID` to understand which "version" of the profile was previously installed on the device.

**Reply by ashishmeher216** [reply]
However, storing one profile with different PayloadUUID  for different devices will be an overhead I believe. What do you think about it?
